### PR TITLE
[12.0] auth_saml: Refactoring using pysaml2

### PR DESCRIPTION
**Pull request by eLBati**

### CODE REVIEW

**Comment by eLBati** [comment]
@theangryangel I'm adding fixes which should be also ported to v13

**Comment by theangryangel** [comment]
Noted. I’m away for a week, I’ll make sure it’s one of the first things I take a look at when im back :+1:

File: auth_saml/models/auth_saml.py [hunk 1]
```diff
@@ -1,78 +1,331 @@
...
+                        ],
+                    },
+                    "allow_unsolicited": self.idp_allow_unsolicited,
+                    "authn_requests_signed": True,
+                    "logout_requests_signed": True,
+                    "want_assertions_signed": True,
+                    # TODO: as we are not correctly generating metadata, don't require
+                    #  signed responses, since default metadata configurations
+                    #  do not require them
+                    "want_response_signed": False,
```
Comment (theangryangel) [inline auth_saml/models/auth_saml.py:173]: Hi @eLBati can you give a bit more detail on this change? It's something I have working in live, if there's a bug with a particular IDP I can see if I can help out? (for ref I've got a copy working in 12.0 against Duo)

**Comment by theangryangel** [comment]
@eLBati FYI I've made some changes to our 13.0 port based on our internal day-to-day usage, it includes a couple of changes;

* Additional (optional) module to filter saml auth providers by website (solves a problem we have where it turns out Duo just ignores the ACS URL it seems, and we need multiple SAML providers configured)
* Fixes the signature issue you pointed out and makes the signature algorithm configurable
* Fixes a couple of issues with with non-fixed base urls
* Allows a user to have multiple SAML providers 

I'll update the 12.0 copy at https://github.com/GlodoUK/server-auth/tree/feat/12.0-mig-auth_saml-pysaml once we've had it running for a few days. I'm not sure if you'll want to update your copy from that

**Comment by eLBati** [comment]
@theangryangel thanks, let me know when you have uploaded the changes

**Comment by ovnicraft** [comment]
@eLBati 
would be nice include controller to response /metadata, what do you think ?

Nice ! https://github.com/OCA/server-auth/pull/234/files#diff-d051ee3f0ce073219b1c7ebc99bfa35206c9ebeefbf62df714fb8c800ad0310cR195

**Comment by moylop260** [comment]
@theangryangel @eLBati 

I just fixed the flake8 checks raised in travis in last commit.

**Comment by moylop260** [comment]
@theangryangel @eLBati
I have changed the requirements.txt in order to install the module correctly in the CI
Pinning the version in order to avoid future errors with incompatible releases

File: auth_saml/models/saml_request.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Copyright (C) 2020 GlodoUK <https://glodo.uk/>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class SamlToken(models.TransientModel):
+    _name = "auth_saml.request"
```
Comment (moylop260) [inline auth_saml/models/saml_request.py:8]: If the provider model is called `auth.saml.provider`

What about calling this one:

```suggestion
    _name = "auth.saml.request"
```

File: auth_saml/__manifest__.py [hunk 3]
```diff
@@ -13,7 +14,7 @@
         'web',
     ],
     "external_dependencies": {
-        "python": ["lasso"],
+        "python": ["saml2"],
     },
     "demo": [
         'demo/auth_saml.xml',
```
Comment (moylop260) [inline auth_saml/__manifest__.py:20]: demo file  show the following warning:

[2020-12-18 04:53:45,018 12075 WARNING auth odoo.models: auth.saml.provider.create() with unknown fields: sp_metadata, sp_pkey](https://travis-ci.com/github/OCA/server-auth/jobs/462174698#L1348)

File: auth_saml/__manifest__.py [hunk 4]
```diff
@@ -13,7 +14,7 @@
         'web',
     ],
     "external_dependencies": {
-        "python": ["lasso"],
```
Comment (yvaucher) [inline auth_saml/__manifest__.py:16]: Dependency is mentionned in the readme, you will want to change it there too.

File: auth_saml/models/auth_saml.py [hunk 5]
```diff
@@ -1,78 +1,331 @@
...
+        detecting from request or system parameter web.base.url
+        """,
     )
-    sp_pkey = fields.Text(
-        string='SP private key',
-        help="Private key for the Service Provider (this Odoo instance)",
+
+    sp_pem = fields.Binary(
+        string="Odoo Certificate",
+        help="Combined certificate and private key used by Odoo to sign",
```
Comment (yvaucher) [inline auth_saml/models/auth_saml.py:51]: Can you state "pem" in the help text to help explaining what kind of file is expected?

File: auth_saml/__manifest__.py (hunk 4 above)
Comment (yvaucher) [reply auth_saml/__manifest__.py:16]: my bad it's in INSTALL.rst, it will be updated automatically

**Review by yvaucher — APPROVED** [review]
Tested locally with a KeyCloak config.

Worked for me to login a user.

**Comment by luisg123v** [comment]
Hi @eLBati,

Is there any news regarding this PR? Do you plan to continue it?

**Comment by theangryangel** [comment]
I can open a refresh PR and incorporate the changes to both the 12.0 and the original 13.0 branch this came from, if there’s interest. I won’t get around to it got a few days tho

**Comment by eLBati** [comment]
Actually we abandoned the project, so I can close this if you want to move to another PR.
Thanks

**Comment by luisg123v** [comment]
@eLBati  yes, please. If not working on this anymore, please close it so the community may now it won't be finished.

**Comment by yvaucher** [comment]
We should definitely use pysaml instead of lasso, it eases the installation of this module.

@eLBati what was the missing part on this that makes you think it's not ready?

**Comment by eLBati** [comment]
@yvaucher just the open remarks of this PR.
We successfully used the module with Azure AD, then the customer moved to other shores

**Comment by ovnicraft** [comment]
@yvaucher module is usable, so i am in favor to open it again an merge it

**Pull request closed without merging**